Documentation/replacing-component.md

```diff
@@ -0,0 +1,266 @@
+struct A: Component {}
+
+// component which implements both `content` and `handle()`
+struct B: Component {
+    var content: some Component { ... }
+    func handle() -> Response { ... }
+}
+```
+
+More generally speaking, fully half of the four possible combinations of conforming to the `Component` protocol will result in invalid components:
+
+<!--- Implement neither `content` nor `handle()`: ❌
+- Implement only `content`: ✅
+- Implement only `handle()`: ✅
+- Implement both `content` and `handle()`: ❌-->
+
+
+|            | `content` | `handle()` | component valid? |
+| :--------- | :-------: | :--------: | :--------------: |
+| implements |  no       | no         | ❌       |
+| implements |  yes      | no         | ✅               |
+| implements |  no       | yes        | ✅               |
+| implements |  yes      | yes        | ❌               |
+
+
+
+
+#### Type inference
+
+Because the current implementation defaults both `Content` and `Response` to `Never`, you can end up in situations where a seemingly properly defined component is in fact ill-formed.
+
+
+Consider the following example:
+
+```swift
+struct User: Codable {
+    let id: ID
+    let name: String
+    let age: Int
+}
+
+struct GetUser: Component {
+    @Parameter("userId")
+    var userId: User.ID
+    
+    func handle() -> User {
+        // ...
+    }
+}
+```
+
+
+This component, despite compiling just fine and not causing a single warning or error, will never work as intended.  
+The reason for this is that `User` doesn't conform to `ResponseEncodable`, meaning that swiftc will resolve `GetUser.Response` to `Never` (the default value), which will cause Apodini to invoke the `handle` function implemented in the `Response == Never` protocol extension.  
+The result is that, when a request is sent to the component's endpoint, it will crash at runtime.
+
+
+**Generally speaking, if we can use the type system to statically detect and reject ill-formed programs, instead of having to perform these checks at runtime, we should always do so.**
+
+
+
+## Proposed replacement
+
+
+### The new protocol structure
+
+My current implementation in [experimental/dsl-node-protocols][branch_url] replaces `Component` with the following two protocols:
+
+```swift
+protocol EndpointNode {
+    associatedtype Response: ResponseEncodable
+    
+    func handle() -> Response
+}
+
+
+protocol EndpointProvidingNode {
+    associatedtype Content: EndpointProvidingNode
+    
+    @EndpointProvidingNodeBuilder
+    var content: Content { get }
+}
+```
+
+- `EndpointNode`s are nodes which expose some functionality to the client.  
+  All leaves in the `WebService` must be `EndpointNode`s.
+- `EndpointProvidingNode`s are nodes which sit above the leaves.  
+  They cannot provide any functionality of their own, but they must eventually lead to one or more `EndpointNode`s.
+
+
+**Note** The names for these protocols aren't fixed, I just went with the most descriptive thing i could think of.
```

Comment:
I would suggest to use the names described in this comment here: https://github.com/Apodini/Apodini/pull/26#discussion_r538396896